src/jackdaw/test/journal.clj

```diff
@@ -76,21 +78,28 @@
     (throw (ex-info "no message stream to journal:" machine)))
 
   (let [{:keys [messages]} (:consumer machine)]
-    (async/go-loop [record (async/<! messages)]
-      (when record
-        (journal-result machine record)
-        (recur (async/<! messages))))))
+    (d/loop [record (s/take! messages)]
+      (d/chain record
+               (fn [record]
+                 (when-not @stop?
```

Comment:
@cddr Doing this you can cancel `d/loop` (stop it) only when there's a new message you can read from the stream. Usually, it's better to check closing conditions periodically on idle. The easiest way would be `take!` with the timeout. E.g.

```clojure
(d/loop [record (s/try-take! message ::drained 1e3 ::timeout)]
  (cond
     (true? @stop)
     <..>

     (identical? ::drained message)
     <..> ;; no more messages

     (identical? ::timeout message)
     (d/recur <..>)))
```